rust_snuba/src/processors/replays.rs

```diff
@@ -0,0 +1,573 @@
+                title,
+                tags_key,
+                tags_value,
+                ..Default::default()
+            };
+
+            vec![serde_json::to_vec(&replay_row)?]
+        }
+        ReplayPayload::EventLinkEvent(event) => {
+            let replay_row = ReplayRow {
+                debug_id: event.debug_id,
+                error_id: event.error_id,
+                event_hash: event.event_hash,
+                fatal_id: event.fatal_id,
+                info_id: event.info_id,
+                offset: metadata.offset,
+                partition: metadata.partition,
+                project_id: replay_message.project_id,
+                replay_id: replay_message.replay_id,
+                retention_days: replay_message.retention_days,
+                timestamp: event.timestamp,
+                warning_id: event.warning_id,
+                ..Default::default()
+            };
+
+            vec![serde_json::to_vec(&replay_row)?]
+        }
+    };
+
+    Ok(InsertBatch {
+        rows: RowData::from_rows(output),
+        origin_timestamp: None,
+        sentry_received_timestamp: None,
+    })
+}
+
+#[derive(Debug, Deserialize)]
+struct ReplayMessage {
+    payload: Vec<u8>,
+    project_id: u64,
+    replay_id: Uuid,
+    retention_days: u16,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type")]
+enum ReplayPayload {
+    #[serde(rename = "replay_actions")]
+    ClickEvent(ReplayClickEvent),
+    #[serde(rename = "replay_event")]
+    Event(Box<ReplayEvent>),
+    #[serde(rename = "event_link")]
+    EventLinkEvent(ReplayEventLinkEvent),
+}
+
+// Replay Click Event
+
+#[derive(Debug, Deserialize)]
+struct ReplayClickEvent {
+    clicks: Vec<ReplayClickEventClick>,
+}
+
+#[derive(Debug, Deserialize)]
+struct ReplayClickEventClick {
+    alt: String,
+    aria_label: String,
+    class: Vec<String>,
+    event_hash: Uuid,
+    id: String,
+    is_dead: u8,
+    is_rage: u8,
+    node_id: u32,
+    role: String,
+    tag: String,
+    testid: String,
+    text: String,
+    timestamp: f64,
+    title: String,
+}
+
+// Replay Event
+
+#[derive(Debug, Deserialize)]
+struct ReplayEvent {
+    #[serde(default)]
+    contexts: Contexts,
+    #[serde(default)]
+    dist: String,
+    #[serde(default)]
+    environment: String,
+    #[serde(default)]
+    event_hash: Option<Uuid>,
+    #[serde(default)]
+    is_archived: u8,
+    #[serde(default = "default_platform")]
+    platform: String,
+    #[serde(default)]
+    release: String,
+    #[serde(default)]
+    replay_start_timestamp: Option<f64>,
+    #[serde(default)]
+    replay_type: String,
+    #[serde(default)]
+    sdk: Version,
+    segment_id: Option<u16>,
+    timestamp: f64,
+    #[serde(default)]
+    urls: Vec<String>,
+    #[serde(default)]
+    user: User,
+    #[serde(default)]
+    trace_ids: Vec<Uuid>,
+    #[serde(default)]
+    error_ids: Vec<Uuid>,
+    #[serde(default)]
+    tags: Vec<Vec<String>>,
```

Comment:
Is this guaranteed to always contain to elements? In which case you can define this as `Vec<(String,String)>` using a tuple.
That should make things more efficient and simplify the downstream code as well.

Reply:
I wasn't aware serde could coerce `[a, b]` to `(a, b)`!  Thank you that's much nicer.